src/main/java/com/example/demo/services/UserService.java

```diff
@@ -0,0 +1,82 @@
+package com.example.demo.services;
+
+import com.example.demo.dtos.ProductDTO;
+import com.example.demo.entities.User;
+import com.example.demo.mappers.DtoMapper;
+import com.example.demo.products.Product;
+import com.example.demo.repositories.UserRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+@Service
+@RequiredArgsConstructor
+public class UserService {
+
+    @Autowired
+    public static UserRepository userRepository;
+    private final DtoMapper mapper;
+
+    public Page<User> pageUsers(Pageable pageRequest){
+        return userRepository.findAll(pageRequest);
+    }
+
+    public void banUser(Long id) {
+        userRepository.findById(id).ifPresent(User -> {
+            User.setActive(false);
+            userRepository.save(User);
+        });
+    }
+
+    public void addProduct(User user, Product product) {
+        user.getProducts().add(product);
+        userRepository.save(user);
+    }
+
+    public void deleteProduct(User user, int index) {
+        user.getProducts().remove(index);
+        userRepository.save(user);
+    }
+
+    public void updateProduct(User user, int index, ProductDTO product) {
+        List<Product> products = user.getProducts();
+        String productType = products.get(index).getClass().getName();
+        switch (productType){
+            case "Ammo" -> products.set(index,mapper.productDtoToAmmo(product));
+            case "AssaultRifle"-> products.set(index,mapper.productDtoToAr(product));
+            case "MachineGun"-> products.set(index,mapper.productDtoToMachinegun(product));
+            case "Pistol"-> products.set(index,mapper.productDtoToPistol(product));
+            case "SniperRifle"-> products.set(index,mapper.productDtoToSniperRifle(product));
+        }
+        userRepository.save(user);
+    }
+
+    public List<? extends Product> otherUserProducts(Long id, String productType) {
+        User user = getUser(id);
+        Stream<Product> productStream = user.getProducts().stream().filter(product -> product.getClass().getName().equals(productType)&&(!product.hidden));
```

Comment:
the chain is too long and unreadable, the filter predicate could be externalized as a method or something.

Also, this is a bad practice to rely on classnames is business logic.

whitespaces around &&

Reply:
I know it's not a good way to rely on class names but I really have no idea how else I can do it. Product types have different attributes, so I can't simply cut them off, and making some Product (super class) attribute which would contain the class names looks redundant while there are inbuilt Java mechanic like that.

Okay, I see one way to do it leaving only Product class - adding enum attribute to it, and depending on that attribute value getters and setters should check if an other attribute in question is allowed/disallowed to get/set for that product type. But isn't it redundant too?